source/common/ssl/ssl_socket.cc

```diff
@@ -195,11 +199,26 @@ Network::IoResult SslSocket::doWrite(Buffer::Instance& write_buffer) {
     }
   }
 
-  return {PostIoAction::KeepOpen, total_bytes_written};
+  if (total_bytes_written == original_buffer_length && last_byte) {
+    shutdownSsl();
+  }
+
+  return {PostIoAction::KeepOpen, total_bytes_written, false};
 }
 
 void SslSocket::onConnected() { ASSERT(!handshake_complete_); }
 
+void SslSocket::shutdownSsl() {
+  if (!shutdown_sent_ && handshake_complete_ &&
```

Comment:
Please comment why we defer until the handshake is complete.  If SSL is shut down and the handshake is incomplete, can it ever complete?  should we flush shutdown when it is?

Edit: oh hey this was moved from prior close.  Well still worth answering IMO :-P

Reply:
Yeah, this is a bit strange, in that connection_impl actually calls close(fd), not the transport, so it's not clear what's happening.  I think this is roughly what is happening (spread across multiple files), but I'm not 100% sure:

```
if (can send ssl shutdown) {
    SSL_shutdown(ssl);
}
close(fd);```